test/e2e/resources_test.go

```diff
@@ -0,0 +1,157 @@
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/knative/serving/test"
+	"k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func isQuotaReached() func(d *v1beta1.Deployment) (bool, error) {
+	return func(d *v1beta1.Deployment) (bool, error) {
+		// TODO Remove this line
+		// fmt.Printf("Conditions %+v\n", d.Status.Conditions)
+		lastCondition := d.Status.Conditions[len(d.Status.Conditions)-1]
+		return lastCondition.Type == v1beta1.DeploymentReplicaFailure &&
+			strings.Contains("quota exceeded", lastCondition.Message), nil
+	}
+}
+
+func satisfyAll(checks ...func(d *v1beta1.Deployment) (bool, error)) func(d *v1beta1.Deployment) (bool, error) {
+	return func(d *v1beta1.Deployment) (bool, error) {
+		for _, check := range checks {
+			ok, err := check(d)
+			if err != nil || !ok {
+				return ok, err
+			}
+		}
+		return true, nil
+	}
+}
+
+func TestUseCustomResources(t *testing.T) {
+	clients := Setup(t)
+
+	//add test case specific name to its own logger
+	logger := logging.GetContextLogger("TestUserCustomResources")
+
+	var imagePath = test.ImagePath("helloworld")
+
+	logger.Infof("Creating a new Route and Configuration")
+	want := corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceMemory: resource.MustParse("666m"),
+			corev1.ResourceCPU:    resource.MustParse("666m"),
+		},
+	}
+	names, err := CreateRouteAndConfig(clients, logger, imagePath, &test.Options{ContainerResources: want})
+	if err != nil {
+		t.Fatalf("Failed to create Route and Configuration: %v", err)
+	}
+	test.CleanupOnInterrupt(func() { TearDown(clients, names, logger) }, logger)
+	defer TearDown(clients, names, logger)
+
+	logger.Infof("When the Revision can have traffic routed to it, the Route is marked as Ready.")
+	if err := test.WaitForRouteState(clients.ServingClient, names.Route, test.IsRouteReady, "RouteIsReady"); err != nil {
+		t.Fatalf("The Route %s was not marked as Ready to serve traffic: %v", names.Route, err)
+	}
+
+	err = test.WaitForPodListState(
+		clients.KubeClient,
+		func(p *v1.PodList) (bool, error) {
+			for _, pod := range p.Items {
+				for _, c := range pod.Spec.Containers {
+					if c.Name == "user-container" {
+						if !equality.Semantic.DeepEqual(c.Resources, want) {
+							return false, fmt.Errorf("Invalid resource configuration. Want: %+v, got: %+v.", want, c.Resources)
+						} else {
+							return true, nil
+						}
+					}
+				}
+			}
+			return false, nil
+		},
+		"WaitForAvailablePods")
+	if err != nil {
+		logger.Fatalf(`Waiting for Pod.List to have pods with custom resources: %v`, err)
+	}
+}
+
+func TestQuotaExceeded(t *testing.T) {
```

Comment:
I don't quite understand how this works. If I wanted to test this for memory, I would:

1. Write a small server (Go, C++ or Python) which uses a POST request parameter to allocates and fills X MB of RAM (you may need to actually touch each memory page to get Linux to allocate the bytes), then frees the memory and reports memory stats in the HTTP response.
1. Set a quota limit of 500MB. Send requests for 100, 200, 800, and make sure that the first two succeed, and the third fails.

Reply:
It makes sense